src/autopas/selectors/tuningStrategy/PredictiveTuning.h

```diff
@@ -0,0 +1,370 @@
+  std::set<Configuration> _optimalSearchSpace;
+
+  /**
+   * Contains the configurations that are not invalid the whole optimalSearchSpace was invalid
+   */
+  std::set<Configuration> _validSearchSpace;
+
+  /**
+   * Gets incremented after every completed tuning phase.
+   * Mainly used for the traversal time storage.
+   */
+  int _timer = 0;
+
+  /**
+   * Checks if a valid configuration was in the optimal SearchSpace.
+   */
+  bool _validConfiguration = false;
+};
+
+void PredictiveTuning::populateSearchSpace(const std::set<ContainerOption> &allowedContainerOptions,
+                                           const std::set<double> &allowedCellSizeFactors,
+                                           const std::set<TraversalOption> &allowedTraversalOptions,
+                                           const std::set<DataLayoutOption> &allowedDataLayoutOptions,
+                                           const std::set<Newton3Option> &allowedNewton3Options) {
+  // generate all potential configs
+  for (auto &containerOption : allowedContainerOptions) {
+    // get all traversals of the container and restrict them to the allowed ones
+    const std::set<TraversalOption> &allContainerTraversals =
+        compatibleTraversals::allCompatibleTraversals(containerOption);
+    std::set<TraversalOption> allowedAndApplicable;
+    std::set_intersection(allowedTraversalOptions.begin(), allowedTraversalOptions.end(),
+                          allContainerTraversals.begin(), allContainerTraversals.end(),
+                          std::inserter(allowedAndApplicable, allowedAndApplicable.begin()));
+
+    for (auto &cellSizeFactor : allowedCellSizeFactors)
+      for (auto &traversalOption : allowedAndApplicable) {
+        for (auto &dataLayoutOption : allowedDataLayoutOptions) {
+          for (auto &newton3Option : allowedNewton3Options) {
+            _searchSpace.emplace(containerOption, cellSizeFactor, traversalOption, dataLayoutOption, newton3Option);
+          }
+        }
+      }
+  }
+
+  AutoPasLog(debug, "Points in search space: {}", _searchSpace.size());
+
+  if (_searchSpace.empty()) {
+    autopas::utils::ExceptionHandler::exception("PredictiveTuning: No valid configurations could be created.");
+  }
+
+  for (auto &configuration : _searchSpace) {
+    std::vector<std::pair<int, size_t>> vector;
+    _traversalTimesStorage.emplace(configuration, vector);
+  }
+
+  _currentConfig = _searchSpace.begin();
+}
+
+void PredictiveTuning::selectPredictedConfigurations() {
+  if (_searchSpace.size() == 1 || _timer == 0 || _timer == 1) {
+    _optimalSearchSpace = _searchSpace;
+    return;
+  }
+
+  predictConfigurations();
+
+  auto optimum = std::min_element(_traversalPredictions.begin(), _traversalPredictions.end(),
+                                  [](std::pair<Configuration, size_t> a, std::pair<Configuration, size_t> b) -> bool {
+                                    return a.second < b.second;
+                                  });
```

Comment:
since you use this `min_element` call at least twice you could write a wrapper for readability like:
```C++
template<class M>
  auto minMapValue(M map) {
    return std::min_element(std::begin(map), std::end(map), [](auto a, auto b) -> bool {
      return a.second < b.second;
    });
  }
```
Maybe you even find a better name than `minMapValue` for better readability.